### PR TITLE
Create more tests on branches and recipes

### PR DESCRIPTION
**Pull request by addyess**
### Overview
* Confirm that branches are available both in launchpad and github for the stable_release, the previous_stable and the current_release
* Raises awareness that stable releases needs a release branch

### Details
* Represents a clean-up/adjustment due to the track map policy changes
* Verifies that upstream stable, upstream previous stable, and this current version all have a stable release branch in Github and Launchpad
* Verify the tip recipes always exist (regardless of stable version)
* GH action must checkout with `depth:0` so it can "see" all the remote branches

### CODE REVIEW

File: .github/workflows/integration.yaml [hunk 1]
```diff
@@ -59,7 +61,7 @@ jobs:
         run: pip install tox
       - name: Run branch_management tests
         run: |
-          tox -c tests/branch_management -e integration
+          tox -c tests/branch_management -e test
```
Comment (addyess) [inline .github/workflows/integration.yaml:64]: ooof!  We weren't running these tests!

File: tests/branch_management/tox.ini [hunk 2]
```diff
@@ -30,12 +30,11 @@ commands =
     black {tox_root}/tests --check --diff
 
 [testenv:test]
-description = Run integration tests
+description = Run branch management tests
 deps =
     -r {tox_root}/requirements-test.txt
 commands =
     pytest -v \
-        --maxfail 1 \
```
Comment (addyess) [inline tests/branch_management/tox.ini:38]: Cmon, let's see all the failures.  Don't stop for just one

File: .github/workflows/integration.yaml [hunk 3]
```diff
@@ -51,6 +51,8 @@ jobs:
     steps:
       - name: Check out code
         uses: actions/checkout@v4
+        with:
+          fetch-depth: 0
```
Comment (addyess) [inline .github/workflows/integration.yaml:55]: lets see all the branches (so we can check that certain branches exist)

File: tests/branch_management/tests/test_branches.py [hunk 4]
```diff
@@ -1,93 +1,127 @@
...
     return [p.name for p in patches]
 
 
-def _confirm_recipe(track, flavour):
+@functools.lru_cache
+def _confirm_branch_exists(repo, branch):
+    log.info(f"Checking {branch} branch exists in {repo}")
+    cmd = f"git ls-remote --heads {repo} {branch}"
+    output = _sh(cmd.split())
+    return branch in output
```
Comment (addyess) [inline tests/branch_management/tests/test_branches.py:42]: the branches shouldn't be changing while we're checking them.  lets cache the results

File: tests/branch_management/tests/test_branches.py [hunk 5]
```diff
@@ -1,93 +1,127 @@
...
+        assert missing, f"GH Branches do not exist {missing}"
+    if missing := [b for b in branches if not _confirm_branch_exists(K8S_LP_REPO, b)]:
+        assert missing, f"LP Branches do not exist {missing}"
+
+
+@functools.lru_cache
+def _confirm_recipe_exist(track, flavour):
     recipe = f"https://launchpad.net/~containers/k8s/+snap/k8s-snap-{track}-{flavour}"
     r = requests.get(recipe)
     return r.status_code == 200
```
Comment (addyess) [inline tests/branch_management/tests/test_branches.py:61]: same for the lp_recipes,

the recipes shouldn't be changing while we're checking them.  lets cache the results

File: tests/branch_management/tests/conftest.py [hunk 6]
```diff
@@ -2,26 +2,71 @@
...
 import requests
 import semver
 
+STABLE_URL = "https://dl.k8s.io/release/stable.txt"
+RELEASE_URL = "https://dl.k8s.io/release/stable-{}.{}.txt"
 
-@pytest.fixture
-def upstream_release() -> semver.VersionInfo:
+
+def _upstream_release_exists(ver: semver.Version) -> Optional[semver.Version]:
```
Comment (bschimke95) [inline tests/branch_management/tests/conftest.py:15]: nit: I'd just name this `_upstream_release`. `_exists` IMHO signals that this function returns a bool

File: tests/branch_management/tests/conftest.py [hunk 7]
```diff
@@ -2,26 +2,71 @@
...
+    """Return true if the major.minor release exists"""
```
Comment (bschimke95) [inline tests/branch_management/tests/conftest.py:16]: This is outdated

File: tests/branch_management/tests/test_branches.py [hunk 8]
```diff
@@ -1,93 +1,127 @@
...
+    ), f"GH Branch {leader} does not exist"
+    branches = [leader]
+    branches += [f"autoupdate/{leader}-{fl}" for fl in _branch_flavours(leader)]
+    if missing := [b for b in branches if not _confirm_branch_exists(K8S_GH_REPO, b)]:
+        assert missing, f"GH Branches do not exist {missing}"
+    if missing := [b for b in branches if not _confirm_branch_exists(K8S_LP_REPO, b)]:
+        assert missing, f"LP Branches do not exist {missing}"
+
+
+@functools.lru_cache
```
Comment (bschimke95) [inline tests/branch_management/tests/test_branches.py:57]: TIL about LRU cache, thanks!

File: tests/branch_management/tests/test_branches.py [hunk 9]
```diff
@@ -1,93 +1,127 @@
...
-    that we do not have a branch for the k8s release for the previous stable release. Let me
-    clarify with an example.
+def test_prior_branches(prior_stable_release):
+    """Ensures git branches exist for prior stable releases.
 
-    Assuming upstream stable k8s release is v1.12.x, there has to be a 1.11 github branch used
-    by the respective LP builders for building the v1.11.y.
+    This is to ensure that the prior release branches exist in the k8s-snap repository
+    before we can proceed to build the next release. For example, if the current stable
+    k8s release is v1.31.0, there must be a release-1.30 branch before updating main.
```
Comment (bschimke95) [inline tests/branch_management/tests/test_branches.py:80]: What happens if we explicitly don't want a release for a specific flavor? Maybe moonray should not have a 1.32 release (for whatever reason). Will this break all releases/checks for this flavor going forward?

**Review by bschimke95 — COMMENTED** [review]
Great work @addyess 
A couple of comments but I like where this is going already.

**Review by bschimke95 — APPROVED** [review]
LGTM!

**Pull request merged**